### ISSUE TITLE
Allow Typing Into Cells Once Highlighted

### ISSUE THREAD

**Issue by seancolsen** [centerofci/mathesar#3236]
### Discussed in https://github.com/centerofci/mathesar/discussions/3233

<div type='discussions-op-text'>

<sup>Originally posted by **MrChadMWood** October  4, 2023</sup>
In the current state, highlighting cells do not make them editable:

Non-Editable:
![image](https://github.com/centerofci/mathesar/assets/90083658/b756a948-e2ad-4b62-b352-8fc6d3ccb2d8)

This makes typing in data manually a little slower, as every cell requires double-click with the mouse.

### Feature Request:
Add the ability (or option) to type into a cell once it is highlighted. 
This way users can navigate across cells while entering data more quickly by -- for instance -- pressing tab, entering the data they want, pressing tab, entering the data they want, ...</div>

**Comment by seancolsen** [comment]
@MrChadMWood One small work-around that may help is the <kbd>Enter</kbd> key &mdash; pressing it converts the cell to edit mode. So your workflow could look like this:

1. <kbd>Enter</kbd>
2. <kbd>Ctrl</kbd>+<kbd>A</kbd> to select all text within the cell
3. Enter new data
4. <kbd>Tab</kbd> or <kbd>Enter</kbd> to save changes

**Comment by seancolsen** [comment]
The behavior that @MrChadMWood proposes here is how I would prefer that Mathesar works &mdash; like a spreadsheet.

I [proposed exactly this behavior](https://github.com/centerofci/mathesar-wiki/pull/13#discussion_r770584517) very early in our UX design process, [making the following case](https://github.com/centerofci/mathesar-wiki/pull/13#discussion_r770870941) for it:

> Given the consistency in UX across different spreadsheet products, I'm a quite wary of intentionally diverging from those well-established patterns. I understand and agree that we need to implement undo before we can safely implement all of the spreadsheet UX patterns. But I don't want to let the sequence of tasks in our roadmap influence our UX design. If we can't do it "the right way" right now, then I'd rather hold off until we _can_ -- instead of implementing a UX pattern that will be unfamiliar to users. I can appreciate that data entry will be more cumbersome if I need to switch to edit mode (via double click or `F2`) before editing a cell. But I think that shortcoming is acceptable while we're still in alpha. For alpha I think it's more important to communicate _the direction the product is heading_ than to supply a product that people can rely on for real-world purposes. If we stray from the spreadsheet UX I think our alpha users will receive some subtle communication from the product that direction is not heading exactly towards spreadsheet-like editing, and that will lower peoples' expectations of our stabilized UX.

However @kgodey disagreed, [making a case](https://github.com/centerofci/mathesar-wiki/pull/13#discussion_r770689503) for adding an extra step:

> I think we should err on the side of not entirely replacing data in the database based on quick actions. We care about data integrity. If the user accidentally types a, then they can remove the a if they make the mistake. If we end up replacing the content of the cell, then they can't recover their previous data, which sounds scary, especially since we don't have undo yet and we've removed it from the plan for the alpha release.

@kgodey later [said](https://github.com/centerofci/mathesar-wiki/pull/13#discussion_r770845614):

> we'll have user feedback to help make the decision.

Now that we have some user feedback, would you be open to reconsidering, @kgodey?

**Comment by ProfessorDisociate** [comment]
Thanks for the explanation, I actually appreciate the effort to preserve integrity.

My use case is but one, however personally I would enjoy the privilege to make these decisions as the admin for my users. My use case is to create a tailored (safe, easy) way for business users to interact with some data that they own. I would like to gauge what “safe” is in this context. 

I like the idea of adding a timer. Personally, I had considered a few options.

Given a sort of cache/queue/buffer is implemented to store unprocessed queries.

There could be a toggle option for “quick mode” that turns on the queue for unprocessed queries.

While the “quick mode” is turned on, there could be a timer. Alternatively, there could be a “COMMIT CHANGES” button somewhere, with a “warning: {n} unsaved changes” next to it.

**Comment by Thanush66** [comment]
Can i work on this..?

**Comment by ghislaineguerin** [comment]
@ProfessorDisociate, thanks for your input. I think your 'quick mode' idea is worth exploring, as it makes sense for users to prioritize speed of entry over safeguarding integrity when necessary.

You've mentioned the need to make other decisions in this context; I'm curious to hear what other ideas you have in mind for admins managing user interactions in Mathesar.

**Comment by ghislaineguerin** [comment]
@seancolsen, as I mentioned in Matrix, collecting use cases involved a lot of data entry across multiple rows, and double-clicking each time was frustrating. Additionally, being able to start typing should also apply to linked record fields, although I assume this could be a separate issue.

**Comment by seancolsen** [comment]
@Thanush66 no, this issue is not ready for implementation yet since it it still `status: draft`. We first need more agreement on our intended behavior. Specifically, we need to hear from @kgodey who had previously voiced concerns over the behavior which this ticket proposes. Then, if we do decide to make this behavior user-configurable, we'll need to do UX design for that configuration functionality first.

**Comment by seancolsen** [comment]
lol turns out I already opened a ticket for this last winter #2381. I closed that as a duplicate, but I do want to paste the "Considerations" I wrote there because they'll be good to keep in mind if we move forward with implementing this at some point:

## Considerations

- We should ensure #1116 is also implemented before cutting a release with this change (to mitigate against inadvertent edits).
- In some types of cells I can only enter certain characters in the input (e.g. Number). What happens when I have selected a Number cell and type a `z` character. Do I enter edit mode? If I do, is the content of the input cleared? Ideally I would _not even enter edit mode_, though I'm not sure how easy that will be to implement. Alternatively, I think it would be okay to enter edit mode but retain the content of the input.
- FK cells do place the user's focus on an input element, but it takes time to fetch data before we render that input element. This is tricky, especially if the user enters many characters rapidly while the record selector is loading. Ideally we'd be able to capture that input while the record selector is loading and transfer it over to the record selector once loaded. It would be especially cool to have an input within the FK cell that shows the user's search query as they type while the record selector is loading (with some kind of loading indicator in the cell too). This seems possible, but probably lots of work. So I think keeping the current behavior of "Enter to launch the record selector" would be fine for now.

**Comment by seancolsen** [comment]
We discussed this issue during today's product approval meeting

Some notes:

- Kriti: ideal solution in my mind would be to have a mode that you can toggle between. Different people might have different needs. On a production DB, you might not want to edit that easily. We should have a default and the support the ability to toggle.
- Pavish: I don't think we should have a toggling state. We should go the "Apple" way — there is one way to do things.
- Kriti: we should be thinking about the problem rather than the solution
- Brent: improving data entry is very important
- Brent: seems like OP didn't know that you could press Enter to enter.

Outcome: this problem is worth solving, we just don't know the solution yet. We need to clarify the requirements to determine which direction we'll head.

**Comment by kgodey** [comment]
Closing, we haven't had any reports of this being an issue and it's still stuck in design discussions. We will open a new, specific implementation issue as needed.
